pyanaconda/modules/storage/bootloader/systemd.py

```diff
@@ -0,0 +1,190 @@
+#
+# Copyright (C) 2022 Arm
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# the GNU General Public License v.2, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY expressed or implied, including the implied warranties of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+#
+#
+#
+#
+# Systemd-boot is dead simple, it basically provides a second boot menu
+# and injects the kernel parms into an efi stubbed kernel, which in turn
+# loads it's initrd. As such there aren't any filesystem or drivers to worry
+# about as everything needed is provided by UEFI. Even the console remains
+# on the UEFI framebuffer, or serial as selected by UEFI. Basically
+# rather than trying to be another mini-os (like grub) and duplicate much of
+# what uefi provides, it simply utilizes the uefi services.
+#
+# Further, while we could keep stage1 (ESP) and stage2 (/boot) seperate it
+# simplifies things just to merge them and place the kernel/initrd on the
+# ESP. This requires a larger than normal ESP, but for now we assume that
+# this linux installer is creating the partitions, so where the space
+# is allocated doesn't matter.
+
+import os
+import re
+
+from pyanaconda.modules.storage.bootloader.base import BootLoader, BootLoaderError
+from pyanaconda.core import util
+from pyanaconda.core.configuration.anaconda import conf
+from pyanaconda.core.i18n import _
+from pyanaconda.core.path import open_with_perm
+from pyanaconda.product import productName
+
+from pyanaconda.anaconda_loggers import get_module_logger
+log = get_module_logger(__name__)
+
+__all__ = ["SYSTEMD"]
+
+class SYSTEMD(BootLoader):
+    """Systemd-boot.
+
+    """
+    name = "SYSTEMD"
+    # oddly systemd-boot files are part of the systemd-udev package
```

Comment:
it's odd indeed. There has been talks with the systemd folks to split it as a separate sub-package.

Reply:
We should get this split out before we proceed any further here.

Reply:
I'm not sure I understand why. There are a half dozen or so pieces needed to get this working, and then about another half dozen cleanups that can/will follow in order to make everything work well. This isn't one of them, as it tosses a few more complexities into the mix because its one of those places where these changes might break something that is already working (aka people already using systemd boot, if the systemd-udev package isn't then made dependent on a systemd-boot package).

Reply:
Because we need Anaconda and other things to consistently understand what package name to request when this is being used. It was _always_ wrong for it to be part of `systemd-udev`. If we don't have this fixed before this landed, we are going to have even more problems down the road.

Reply:
? But that only matters during installation.  Along with the bootctl install flags/location, and a few other things, which presumably _might_ change in the future, it should be fairly easy to update anaconda if/when something like that changes, as it shouldn't get out of sync. 

Maybe I"m misunderstanding. 

Either way, you are probably right that the sooner the switchover in systemd happens (if its to happen at all), the better in order to avoid the problem I noted above.

Reply:
Well, given there are a bunch of systemd-boot PR's lying about, I added one of my own, slightly different in that it moves both the userspace and the UEFI executables into the same package.

https://src.fedoraproject.org/rpms/systemd/pull-request/95

Reply:
Oh, I understand what happened now. In the PR for systemd there was no mention that is related to Anaconda. The PR was doing something very close to the older PR, so I merged that one instead.

> oddly systemd-boot files are part of the systemd-udev package

Systemds has a bazillion little helpers, and if we tried to split them up by functionality, we'd end up with many tiny packages. Over the years there were a few attempts to split things up, and in the end we ended up with the current form where things are split when there is a benefit. So for example, systemd-udev.rpm has udev, rfkill, various sleep tools, timesyncd, bintfmt, fsck, and even coredump. The reason is that all those things would generally be needed on any standalone machine (either physical or vm), but not in a container. The main systemd package contains anything to boot a machine and maintain it (physical, vm, or container). systemd-journal-remote.rpm contains tools to send journal files over the network, and pulls in additional http and encryption libraries, which we want to avoid by default, hence the separate package. systemd-pam, systemd-libs, systemd-rpm-macros are also separate to not require full systemd installation in various scenarios.

And as mentioned before, various `*ctl` utilities are often in systemd.rpm. They work over the network and on containers, so they are useful without the service part being installed. And they often are useful to query if the service is actually used. Both coredumpctl and bootctl only work locally and wouldn't be useful in non-machine systems.

There's also some work in progress to add support in grub2 to the Boot Loader Interface, so bootctl will be useful even without sd-boot.

None of this is set in stone, and I'm of course happy to discuss changes, but there are reasons for the current state.

Reply:
The real problem isn't this line, so much as where anaconda updates the dependency lists for the install target

https://github.com/jlinton/anaconda/blob/allow_systemd_boot_as_bootloader/pyanaconda/modules/storage/bootloader/efi.py#L206

Without a stable package name (across signed/unsigned/whatever) systemd-boot utilities+bootloaders the bootctl and EFI utilities won't be installed correctly.

Reply:
https://src.fedoraproject.org/rpms/systemd/c/732bdcb223ae95b41e37aa1ba1a3256781a51fbd?branch=rawhide adds
```
Provides:systemd-boot = …
Provides:systemd-boot%{_isa} = …
```

Reply:
I'm going to close this one out because I think the main complaints have been addressed, and the latest push adds the systemd-boot dependency.